

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>immuno_probs.cdr3 package &mdash; ImmunoProbs 0.2.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="immuno_probs.cli package" href="immuno_probs.cli.html" />
    <link rel="prev" title="immuno_probs.alignment package" href="immuno_probs.alignment.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> ImmunoProbs
          

          
          </a>

          
            
            
              <div class="version">
                0.2.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Readme</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../use_cases.html">Use cases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../models.html">Models</a></li>
</ul>
<p class="caption"><span class="caption-text">Modules</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">immuno_probs</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="immuno_probs.html">immuno_probs package</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="immuno_probs.alignment.html">immuno_probs.alignment package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">immuno_probs.cdr3 package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-immuno_probs.cdr3.anchor_locator">immuno_probs.cdr3.anchor_locator module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-immuno_probs.cdr3.olga_container">immuno_probs.cdr3.olga_container module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="immuno_probs.cli.html">immuno_probs.cli package</a></li>
<li class="toctree-l3"><a class="reference internal" href="immuno_probs.convert.html">immuno_probs.convert package</a></li>
<li class="toctree-l3"><a class="reference internal" href="immuno_probs.model.html">immuno_probs.model package</a></li>
<li class="toctree-l3"><a class="reference internal" href="immuno_probs.util.html">immuno_probs.util package</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ImmunoProbs</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="modules.html">immuno_probs</a> &raquo;</li>
        
          <li><a href="immuno_probs.html">immuno_probs package</a> &raquo;</li>
        
      <li>immuno_probs.cdr3 package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="immuno-probs-cdr3-package">
<h1>immuno_probs.cdr3 package<a class="headerlink" href="#immuno-probs-cdr3-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-immuno_probs.cdr3.anchor_locator">
<span id="immuno-probs-cdr3-anchor-locator-module"></span><h2>immuno_probs.cdr3.anchor_locator module<a class="headerlink" href="#module-immuno_probs.cdr3.anchor_locator" title="Permalink to this headline">¶</a></h2>
<p>Contains AnchorLocator class for locating CDR3 anchors of given sequences.</p>
<dl class="class">
<dt id="immuno_probs.cdr3.anchor_locator.AnchorLocator">
<em class="property">class </em><code class="descclassname">immuno_probs.cdr3.anchor_locator.</code><code class="descname">AnchorLocator</code><span class="sig-paren">(</span><em>alignment</em>, <em>gene</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/immuno_probs/cdr3/anchor_locator.html#AnchorLocator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#immuno_probs.cdr3.anchor_locator.AnchorLocator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Locates all possibile CDR3 start and end positions within the given nucleotide alignment.</p>
<p>This class followes OLGA’s formatting standard for the outputed CDR3 anchor files. The gene names in the reference genomic
sequence alignment should be located in the second position and the function of the gene on the fourth position in the
sequence header (separated by ‘|’). The header needs to follow IMGT standards.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>alignment</strong> (<em>Bio.AlignIO</em>) – An biopython MUSCLE alignement object created with the alignment.MuscleAligner class.</li>
<li><strong>gene</strong> (<em>str</em>) – A gene identifier, either V or J, specifying the alignment’s origin gene.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="immuno_probs.cdr3.anchor_locator.AnchorLocator.get_indices_motifs">
<code class="descname">get_indices_motifs</code><span class="sig-paren">(</span><em>num_threads</em>, <em>*motifs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/immuno_probs/cdr3/anchor_locator.html#AnchorLocator.get_indices_motifs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#immuno_probs.cdr3.anchor_locator.AnchorLocator.get_indices_motifs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a pandas.DataFrame containing CDR3 anchors.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">get_indices_motifs</code><span class="sig-paren">(</span><em>num_threads</em>, <em>*motifs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/immuno_probs/cdr3/anchor_locator.html#AnchorLocator.get_indices_motifs"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Collects and returns the CDR3 anchors for each motif from the sequence alignment.</p>
<p>The function locates the most common V (Cysteine - TGT and TGC by default) or J (Tryptophan - TGG,
Phenylalanine - TTC and TTT by default) gene index that best covers all sequences in the alignment.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>num_threads</strong> (<em>int</em>) – The number of threads to use when processing the motif list.</li>
<li><strong>*motifs</strong> (<em>str</em>) – Various motif strings (combined into a list) to process. At least motif needs to be specified.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Containing columns with sequence identifier, start index values for the anchors, function of the gene and motifs.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">pandas.DataFrame</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – When no motifs have been specified in the function call.</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>This function uses the given MUSCLE alignment and gene identifier from the class constructor.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-immuno_probs.cdr3.olga_container">
<span id="immuno-probs-cdr3-olga-container-module"></span><h2>immuno_probs.cdr3.olga_container module<a class="headerlink" href="#module-immuno_probs.cdr3.olga_container" title="Permalink to this headline">¶</a></h2>
<p>Contains OlgaContainer class for generating and evaluating CDR3 sequences.</p>
<dl class="class">
<dt id="immuno_probs.cdr3.olga_container.OlgaContainer">
<em class="property">class </em><code class="descclassname">immuno_probs.cdr3.olga_container.</code><code class="descname">OlgaContainer</code><span class="sig-paren">(</span><em>igor_model</em>, <em>nt_col</em>, <em>nt_p_col</em>, <em>aa_col</em>, <em>aa_p_col</em>, <em>v_gene_choice_col</em>, <em>j_gene_choice_col</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/immuno_probs/cdr3/olga_container.html#OlgaContainer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#immuno_probs.cdr3.olga_container.OlgaContainer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Generates and/or evaluates CDR3 sequences using given an IGoR model.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>igor_model</strong> (<em>immuno_probs.model.igor_loader.IgorLoader</em>) – IgorLoader object containing the loaded IGoR VJ or VDJ model.</li>
<li><strong>nt_col</strong> (<em>str</em>) – The name of the nucleotide sequence column to use.</li>
<li><strong>nt_p_col</strong> (<em>str</em>) – The name of the nucleotide Pgen column to use.</li>
<li><strong>aa_col</strong> (<em>str</em>) – The name of the aminoacid sequence column to use.</li>
<li><strong>aa_p_col</strong> (<em>str</em>) – The name of the aminoacid Pgen column to use.</li>
<li><strong>v_gene_choice_col</strong> (<em>str</em>) – The name of the V gene choice column to use.</li>
<li><strong>j_gene_choice_col</strong> (<em>str</em>) – The name of the J gene choice column to use.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="immuno_probs.cdr3.olga_container.OlgaContainer.generate">
<code class="descname">generate</code><span class="sig-paren">(</span><em>num_seqs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/immuno_probs/cdr3/olga_container.html#OlgaContainer.generate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#immuno_probs.cdr3.olga_container.OlgaContainer.generate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns pandas.DataFrame with nucleotide and aminoacid CDR3 sequences.</p>
</dd></dl>

<dl class="method">
<dt id="immuno_probs.cdr3.olga_container.OlgaContainer.evaluate">
<code class="descname">evaluate</code><span class="sig-paren">(</span><em>seq</em>, <em>num_threads</em>, <em>use_allele=True</em>, <em>default_allele=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/immuno_probs/cdr3/olga_container.html#OlgaContainer.evaluate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#immuno_probs.cdr3.olga_container.OlgaContainer.evaluate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the generation probability value for the given sequences.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">evaluate</code><span class="sig-paren">(</span><em>seqs</em>, <em>num_threads</em>, <em>use_allele=True</em>, <em>default_allele=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/immuno_probs/cdr3/olga_container.html#OlgaContainer.evaluate"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Evaluate a given nucleotide CDR3 sequences using OLGA.</p>
<p>This function also checks if the given input sequence file contains the gene index columns for the V and J gene.
If so, then the V and J gene masks in these columns are used to increase calculation accuracy of the generation
probabality values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seqs</strong> (<em>pandas.DataFrame</em>) – A pandas dataframe object containing a column with nucleotide CDR3 sequences and/or amino acid sequences.</li>
<li><strong>num_threads</strong> (<em>int</em>) – The number of threads to use when processing the sequences.</li>
<li><strong>use_allele</strong> (<em>bool, optional</em>) – If True, the allele information from the input genes is used instead of the ‘default_allele’ value (default: True).</li>
<li><strong>default_allele</strong> (<em>str, optional</em>) – A default allele value to use when spliting gene choices, and ‘use_allele’ option is False (default: None).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Containing the sequence index number, the generation probability of nucleotide sequence if given and the
generation probability of aminoacid sequence if given.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">pandas.DataFrame</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.8)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">TypeError</span></code></a> – When the model type does not equal ‘VDJ’ or ‘VJ’.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">generate</code><span class="sig-paren">(</span><em>num_seqs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/immuno_probs/cdr3/olga_container.html#OlgaContainer.generate"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Generate a given number of CDR3 sequences through OLGA.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>num_seqs</strong> (<em>int</em>) – An integer specifying the number of sequences to generate.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Containing columns with sequence index, nucleotide CDR3 sequence, amino acid CDR3 sequence, the index of the
chosen V gene and the index of the chosen J gene.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">pandas.DataFrame</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.8)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">TypeError</span></code></a> – When the model type does not equal ‘VDJ’ or ‘VJ’.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="immuno_probs.cli.html" class="btn btn-neutral float-right" title="immuno_probs.cli package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="immuno_probs.alignment.html" class="btn btn-neutral float-left" title="immuno_probs.alignment package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Wout van Helvoirt

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>